http-common/src/connector.rs

```diff
@@ -72,36 +68,174 @@ impl Incoming {
-                                limit_err
-                            );
-                        }
-                    }
+                    serve_unix(user_state, server, unix_stream).await;
                 });
             },
         }
     }
+
+    #[cfg(feature = "serve_with_shutdown")]
+    pub async fn serve_with_shutdown<H>(
+        &mut self,
+        server: H,
+        shutdown: tokio::sync::oneshot::Receiver<()>,
+    ) -> std::io::Result<()>
+    where
+        H: hyper::service::Service<
+                hyper::Request<hyper::Body>,
+                Response = hyper::Response<hyper::Body>,
+                Error = std::convert::Infallible,
+            > + Clone
+            + Send
+            + 'static,
+        <H as hyper::service::Service<hyper::Request<hyper::Body>>>::Future: Send,
+    {
+        // Keep track of the number of running tasks.
+        let tasks = std::sync::atomic::AtomicIsize::new(0);
+        let tasks = std::sync::Arc::new(tasks);
+
+        let shutdown_loop = shutdown;
+        futures_util::pin_mut!(shutdown_loop);
+
+        match self {
+            Incoming::Tcp { listener } => loop {
+                let accept = listener.accept();
+                futures_util::pin_mut!(accept);
+
+                match future::select(shutdown_loop, accept).await {
+                    future::Either::Left((_, _)) => break,
+                    future::Either::Right((tcp_stream, shutdown)) => {
+                        let tcp_stream = tcp_stream?.0;
+
+                        let server = crate::uid::UidService::new(None, 0, server.clone());
+
+                        tasks.fetch_add(1, std::sync::atomic::Ordering::Relaxed);
+                        let server_tasks = tasks.clone();
+                        tokio::spawn(async move {
+                            serve_tcp(server, tcp_stream).await;
+                            server_tasks.fetch_sub(1, std::sync::atomic::Ordering::Relaxed);
+                        });
+
+                        shutdown_loop = shutdown;
+                    }
+                }
+            },
+
+            Incoming::Unix {
+                listener,
+                user_state,
+            } => loop {
+                let accept = listener.accept();
+                futures_util::pin_mut!(accept);
+
+                // Await either the next established connection or the shutdown signal.
+                match future::select(shutdown_loop, accept).await {
+                    future::Either::Left((_, _)) => break,
+                    future::Either::Right((unix_stream, shutdown)) => {
+                        let unix_stream = unix_stream?.0;
+
+                        let (user_state, ucred) = get_user_auth(&unix_stream, user_state)?;
+                        let server =
+                            crate::uid::UidService::new(ucred.pid(), ucred.uid(), server.clone());
+
+                        tasks.fetch_add(1, std::sync::atomic::Ordering::Relaxed);
```

Comment:
I don't think `Relaxed` is appropriate for any of the three. The two `fetch_`s should be `AcqRel` and the `load` should be `Acquire`. `Relaxed` allows them to move around, so for example the two `fetch_`s would not necessarily happen before and after the `serve_unix` call.

Reply:
Fixed.